src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -4279,18 +4385,60 @@ private static BoundExpression SkipReferenceConversions(BoundExpression possibly
                 // In the when-null branch, the receiver is known to be maybe-null.
                 // In the other branch, the receiver is known to be non-null.
                 LearnFromNullTest(receiver, ref receiverState);
-                LearnFromNonNullTest(receiver, ref this.State);
-                var nextConditionalAccessSlot = MakeSlot(receiver);
-                if (nextConditionalAccessSlot > 0 && receiver.Type?.IsNullableType() == true)
-                    nextConditionalAccessSlot = GetNullableOfTValueSlot(receiver.Type, nextConditionalAccessSlot, out _);
+                makeAndAdjustReceiverSlot(receiver);
+            }
 
-                _lastConditionalAccessSlot = nextConditionalAccessSlot;
+            // We want to preserve stateWhenNotNull from accesses in the same "chain":
+            // a?.b(out x)?.c(out y); // expected to preserve stateWhenNotNull from both ?.b(out x) and ?.c(out y)
+            // but not accesses in nested expressions:
+            // a?.b(out x, c?.d(out y)); // expected to preserve stateWhenNotNull from a?.b(out x, ...) but not from c?.d(out y)
+            BoundExpression expr = node.AccessExpression;
+            while (expr is BoundConditionalAccess innerCondAccess)
+            {
+                // we assume that non-conditional accesses can never contain conditional accesses from the same "chain".
+                // that is, we never have to dig through non-conditional accesses to find and handle conditional accesses.
+                VisitRvalue(innerCondAccess.Receiver);
+                _currentConditionalReceiverVisitResult = _visitResult;
+                makeAndAdjustReceiverSlot(innerCondAccess.Receiver);
+
+                // The receiverState here represents the scenario where 0 or more of the access expressions could have been evaluated.
+                // e.g. after visiting `a?.b(x = null)?.c(x = new object())`, the "state when not null" of `x` is NotNull, but the "state when maybe null" of `x` is MaybeNull.
+                Join(ref receiverState, ref State);
+
+                expr = innerCondAccess.AccessExpression;
             }
 
-            var accessTypeWithAnnotations = VisitLvalueWithAnnotations(node.AccessExpression);
-            TypeSymbol accessType = accessTypeWithAnnotations.Type;
+            Debug.Assert(expr is BoundExpression);
+            Visit(expr);
+
+            expr = node.AccessExpression;
+            while (expr is BoundConditionalAccess innerCondAccess)
+            {
+                // The resulting nullability of each nested conditional access is the same as the resulting nullability of the rightmost access.
```

Comment:
That doesn't seem obvious. I would have expected we'd record the state/result as we computed the state after each access (in the `while` loop above).

Reply:
This is more intuitive when the visit is implemented recursively.

Consider a hypothetical expression `a?.b?.c`:
```
Visit // a?.b?.c
    VisitReceiver // a
    Visit // b?.c
        VisitReceiver // ('a' conditional receiver).b
        Visit // c
        SetResult
    SetResult
SetResult
```

They're all setting the same result based on what they got from the rightmost access expression.

Reply:
Ah, indeed. Thanks

---
In reply to: [620720440](https://github.com/dotnet/roslyn/pull/52646#discussion_r620720440) [](ancestors = 620720440)